cmd/helm/push_test.go

```diff
@@ -17,11 +17,138 @@ limitations under the License.
 package main
 
 import (
+	"bytes"
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"net"
+	"os"
 	"testing"
+	"time"
+
+	"helm.sh/helm/v3/pkg/action"
+	helmRegistry "helm.sh/helm/v3/pkg/registry"
+
+	"github.com/distribution/distribution/v3/configuration"
+	"github.com/distribution/distribution/v3/registry"
+	"github.com/phayes/freeport"
+	"github.com/stretchr/testify/suite"
 )
 
 func TestPushFileCompletion(t *testing.T) {
 	checkFileCompletion(t, "push", true)
 	checkFileCompletion(t, "push package.tgz", false)
 	checkFileCompletion(t, "push package.tgz oci://localhost:5000", false)
 }
+
+var (
+	testWorkspaceDir = "helm-registry-test"
+)
+
+type HelmPushTestSuite struct {
+	suite.Suite
+	Out                     io.Writer
+	DockerRegistryHost      string
+	CompromisedRegistryHost string
+	WorkspaceDir            string
+	RegistryClient          *helmRegistry.Client
+	PushClient              *action.Push
+}
+
+func getClientIP() (string, error) {
+	addrs, err := net.InterfaceAddrs()
+
+	if err != nil {
+		return "", err
+	}
+	for _, address := range addrs {
+		// filter loopback ip
+		if ipNet, ok := address.(*net.IPNet); ok && !ipNet.IP.IsLoopback() {
+			if ipNet.IP.To4() != nil {
+				return ipNet.IP.String(), nil
+			}
+		}
+	}
+
+	return "", errors.New("can not find the client ip address")
+
+}
+
+func (suite *HelmPushTestSuite) SetupSuite() {
+	suite.WorkspaceDir = testWorkspaceDir
+	os.RemoveAll(suite.WorkspaceDir)
+	os.Mkdir(suite.WorkspaceDir, 0700)
+
+	var out bytes.Buffer
+	suite.Out = &out
+	//credentialsFile := filepath.Join(suite.WorkspaceDir, registry.CredentialsFileBasename)
+
+	// Registry config
+	config := &configuration.Configuration{}
+	port, err := freeport.GetFreePort()
+	suite.Nil(err, "no error finding free port for test registry")
+
+	ip, err := getClientIP()
+	suite.Nil(err, "get host ip error")
+	suite.DockerRegistryHost = fmt.Sprintf("%s:%d", ip, port)
+	config.HTTP.Addr = fmt.Sprintf("%s:%d", ip, port)
+	config.HTTP.DrainTimeout = time.Duration(10) * time.Second
+	config.Storage = map[string]configuration.Parameters{"inmemory": map[string]interface{}{}}
+
+	dockerRegistry, err := registry.NewRegistry(context.Background(), config)
+	suite.Nil(err, "no error creating test registry")
+
+	suite.CompromisedRegistryHost = fmt.Sprintf("%s:%d", ip, port)
+
+	// Start Docker registry
+	go dockerRegistry.ListenAndServe()
+}
+
+func (suite *HelmPushTestSuite) Test_1_Push() {
```

Comment:
try to write test like:
```golang
tests:=[]struct{}
for _,tt := range tests {}
```
Thanks very much

Reply:
I think I can make the test more intuitive by changing the function name. For example, `Test_1_Push` is changed to `TestSecretPush`, and `Test_2_Push` is changed to `TestInsecretPush`. Currently, each test function has initial logic, and it is not easy to initialize using this format.

Reply:
ok